source /groups/pupko/yairshimony/miniconda3/etc/profile.d/conda.sh
conda activate /groups/pupko/yairshimony/miniconda3/envs/microbializer
export PATH=$CONDA_PREFIX/bin:$PATH
python /groups/pupko/yairshimony/microbializer/pipeline/steps/extract_core_genome.py /groups/pupko/yairshimony/microbializer_runs/1_genome_small/outputs/steps_results/08_1_orthogroups_fasta/orthogroups_aa_msa /groups/pupko/yairshimony/microbializer_runs/1_genome_small/outputs/genomes_names.txt /groups/pupko/yairshimony/microbializer_runs/1_genome_small/outputs/steps_results/09_3_aligned_core_proteome_reduced/aligned_core_proteome.fas /groups/pupko/yairshimony/microbializer_runs/1_genome_small/outputs/steps_results/09_3_aligned_core_proteome_reduced/core_length.txt --core_minimal_percentage 100.0 --max_number_of_ogs 1000 -v False --logs_dir /groups/pupko/yairshimony/microbializer_runs/1_genome_small/outputs/tmp/09_3_aligned_core_proteome_reduced --error_file_path /groups/pupko/yairshimony/microbializer_runs/1_genome_small/M1CR0B1AL1Z3R_outputs/error.txt --job_name core_proteome --use_job_manager True --cpus 1
touch /groups/pupko/yairshimony/microbializer_runs/1_genome_small/outputs/tmp/09_3_aligned_core_proteome_reduced/core_proteome.done
